### PR TITLE
Revisit and improve checkers hands-on-exercise

### PR DESCRIPTION
**Pull request by xavierlepretre**
Documentation content update for the hands-on-exercise.

This PR revisits and improves certain aspects of the checkers hands-on-exercise

The file numbering and the in-file ordering are incorrect. This would have to be fixed with a quick PR after merge.

### R1: Internal review (B9lab)

- Technical review
  - [x] requested
  - [ ] completed
- Language review
  - [x] requested
  - [x] completed
- Technical platform review (remove if not required)
  - [x] requested
  - [ ] completed

### R2: External review

- Technical review
  - [x] requested
  - [x] completed
- Language review
  - [x] requested
  - [x] completed

### R3: Internal QA review (B9lab)

- Technical review
  - [ ] requested
  - [ ] completed
- Language review
  - [ ] requested
  - [ ] completed

### R4: Final external QA review

- Technical review
  - [ ] requested
  - [ ] completed
- Language review
  - [ ] requested
  - [ ] completed

### RC: Release Candidate

- [ ] Ready to be merged

### CODE REVIEW

File: hands-on-exercise/2-ignite-cli-adv/6-gas-meter.md [hunk 1]
```diff
@@ -89,30 +89,27 @@ Add a line that consumes or refunds the designated amount of gas in each relevan
...
-    ```diff-go [https://github.com/cosmos/b9-checkers-academy-draft/blob/gas-meter/x/checkers/keeper/msg_server_play_move.go#L88]
+    ```diff-go [https://github.com/cosmos/b9-checkers-academy-draft/blob/gas-meter/x/checkers/keeper/msg_server_play_move.go#L90]
         ...
         k.Keeper.SetSystemInfo(ctx, systemInfo)
     +  ctx.GasMeter().ConsumeGas(types.PlayMoveGas, "Play a move")
         ...
     ```
 
-3. When handling a game rejection, you make sure that you are not refunding more than what has already been consumed:
+Note that when handling a move, the gas is measured towards the end. This is a design decision, and you may arrive at a different conclusion. Indeed, if a player hits an error, such as a failure to pay the wager, or passing an illegal move, then the code will return an error before the gas metering line. This means this _errored_ player will pay only the gas already metered by the rest of the Cosmos SDK. If you decide to charge extra gas for a move whatever the errors, then you would move the gas metering line closer to the beginning of the function's body.
```
Comment (AndrewB9lab) [inline hands-on-exercise/2-ignite-cli-adv/6-gas-meter.md:99]: @xavierlepretre 
> This means this _errored_ player will pay only the gas already metered by the rest of the Cosmos SDK.

This refers to _the player **making** the error_,  correct?

Comment (xavierlepretre) [reply hands-on-exercise/2-ignite-cli-adv/6-gas-meter.md:99]: > This refers to _the player **making** the error_, correct?

Yes

Comment (xavierlepretre) [reply hands-on-exercise/2-ignite-cli-adv/6-gas-meter.md:99]: > This means, in effect, "charging gas for moves that are not accepted due to errors", yes?

Both our versions are too confusing. The true / false pivot here is:

> the gas is measured towards the end

* If this is true (as currently implemented), then -> This means this _errored_ player will pay only the gas already metered by the rest of the Cosmos SDK
* If this is false (as is up to the developer), then the gas would be measure towards the beginning of the function, then that's equivalent to -> you decide to charge extra gas for a move whatever the errors

Comment (AndrewB9lab) [reply hands-on-exercise/2-ignite-cli-adv/6-gas-meter.md:99]: Thanks for this :-) my review push will include a revision on this point, so I'll add a note then to draw attention.

File: hands-on-exercise/3-cosmjs-adv/2-cosmjs-messages.md [hunk 2]
```diff
@@ -577,16 +610,16 @@ Fortunately, the [`sign`](https://github.com/cosmos/cosmjs/blob/v0.28.11/package
...
 
 On the other hand, you want the last transaction to be included in the block so that when you query for the stored game you get the expected values. Therefore you will:
 
 1. Send the first 21 signed transactions with the _fast_ `this.forceGetTmClient().broadcastTxSync`.
 2. Send the last transaction with a _slow_ client `broadcastTx`.
 
-Here again, you need to make sure that you submit all transactions in sequential manner, otherwise a player may in effect try to play before their turn. At this point, you trust that Tendermint includes the transactions in the order in which they were submitted. If Tendermint does any shuffling between Alice and Bob, you may end up with a "play before their turn" error.
+Here again, you need to make sure that you submit all transactions in sequential manner, otherwise a player may in effect try to play before their turn. At this point, you have to trust that Tendermint includes the transactions in the order in which they were submitted. If Tendermint does any shuffling between Alice and Bob, you may end up with a "play before their turn" error.
```
Comment (AndrewB9lab) [inline hands-on-exercise/3-cosmjs-adv/2-cosmjs-messages.md:622]: @xavierlepretre we have references to "Tendermint" here. Are we discussing _Tendermint-the-algorithm_, or should these be changed to read _CometBFT_?

My sense is that we are discussing _the functionality of the tool providing Consensus_, not the algorithm it incorporates, so I should switch to _CometBFT_. But I await your confirmation!

File: hands-on-exercise/1-ignite-cli/6-play-game.md [hunk 3]
```diff
@@ -180,7 +180,7 @@ Then you can check that:
     }
     ```
 
-2. The positions are within bounds. With a array of situations:
+2. The positions are within bounds, checking a array of situations:
```
Comment (xavierlepretre) [inline hands-on-exercise/1-ignite-cli/6-play-game.md:183]:
```suggestion
2. The positions are within bounds, checking an array of situations:
```

File: hands-on-exercise/4-run-in-prod/1-run-prod-docker.md [hunk 4]
```diff
@@ -1011,6 +1157,99 @@ Note that it has to replace only the first occurrence in the whole.
...
+      checkersd_i \
+      -Ei 's/^enabled-unsafe-cors = false/enabled-unsafe-cors = true/g' \
+      /root/.checkers/config/app.toml
+  ```
+
+  </CodeGroupItem>
+
+  </CodeGroup>
+
+1. In `app.toml`, first location:
```
Comment (AndrewB9lab) [inline hands-on-exercise/4-run-in-prod/1-run-prod-docker.md:1223]: @xavierlepretre This repeats the string from Line 1193. Should it read "second location" here?

File: hands-on-exercise/2-ignite-cli-adv/6-gas-meter.md [hunk 5]
```diff
@@ -96,15 +96,29 @@ Add a line that consumes or refunds the designated amount of gas in each relevan
...
+In the current arrangement, if a player hits an error (such as a failure to pay the wager, or making an illegal move) then the code will return an error _before_ the gas metering line. This means the player making the error will only pay the gas already metered by the rest of the Cosmos SDK.
+
+If you decide to move the gas metering line closer to the beginning of the function's body, then you will charge the player extra gas for any move that contains errors _despite the fact that the requested move will not be accepted_.
+
+This alternate approach would certainly dis-incentivize players from submitting erroneous moves, but it also risks alienating them through financial punishment for accidental mistakes. Better is to implement a feature that allows players to check a move is valid _before_ they pay the cost of handling. This is explored in the [next section](/hands-on-exercise/2-ignite-cli-adv/7-can-play.md).
+
+</HighlightBox>
+
+You are free to consider charging different gas costs for when a player _plays a regular legal move_ and for when a player _plays a winning move_. After a win the board is cleared, and you therefore may want to exact a smaller cost in this situation.
```
Comment (xavierlepretre) [inline hands-on-exercise/2-ignite-cli-adv/6-gas-meter.md:111]:
```suggestion
You are free to consider charging different gas costs for when a player _plays a regular legal move_ and for when a player _plays a winning move_. After a win the board is cleared, and you therefore may want to exact a smaller cost (or a gas refund even) in this situation.
```

File: hands-on-exercise/4-run-in-prod/1-run-prod-docker.md (hunk 4 above)
Comment (xavierlepretre) [reply hands-on-exercise/4-run-in-prod/1-run-prod-docker.md:1223]: Oops yes, it is "second location".

**Review by AndrewB9lab — APPROVED** [review]
Language review on the checkers improvements is completed.

**Review by murataniloener — APPROVED** [review]
lgtm, great work!

**Review by tmsdkeys — APPROVED** [review]
Reviewed these sections:

- academy/3-ibc
- hands-on-exercise
   - 1-iginite-cli
   -  3-cosmjs-adv
   - 5-ibc-adv
Left few nits, but LGTM for the most part. Nice cleanup!

**Review by samricotta — APPROVED** [review]

**Review by coldice — APPROVED** [review]

**Pull request merged**